doc/sphinx/proof-engine/tactics.rst

```diff
@@ -898,314 +1113,235 @@ Applying theorems
-   type of :token:`ident` against a non-dependent premise of the type
-   of :token:`term`, trying them from right to left. If it succeeds, the
-   statement of hypothesis :token:`ident` is replaced by the conclusion of
-   the type of :token:`term`. The tactic also returns as many subgoals as the
-   number of other non-dependent premises in the type of :token:`term` and of
-   the non-dependent premises of the type of :token:`ident`. If the conclusion
-   of the type of :token:`term` does not match the goal *and* the conclusion
-   is an inductive type isomorphic to a tuple type, then
-   the tuple is (recursively) decomposed and the first component of the tuple
-   of which a non-dependent premise matches the conclusion of the type of
-   :token:`ident`. Tuples are decomposed in a width-first left-to-right order
-   (for instance if the type of :g:`H1` is :g:`A <-> B` and the type of
-   :g:`H2` is :g:`A` then :g:`apply H1 in H2` transforms the type of :g:`H2`
-   into :g:`B`). The tactic :tacn:`apply` relies on first-order pattern matching
-   with dependent types.
-
-   .. exn:: Statement without assumptions.
-
-      This happens if the type of :token:`term` has no non-dependent premise.
-
-   .. exn:: Unable to apply.
+.. _managingthelocalcontext:
 
-      This happens if the conclusion of :token:`ident` does not match any of
-      the non-dependent premises of the type of :token:`term`.
+Managing the local context
+------------------------------
 
-   .. tacv:: apply {+, @term} in {+, @ident}
+.. tacn:: intro {? @ident } {? @where }
 
-      This applies each :token:`term` in sequence in each hypothesis :token:`ident`.
+   Applies the :tacn:`hnf` tactic until it finds an item that can be
+   introduced in the context by removing certain constructs in the goal.
+   If no item is found, the tactic fails.  The name used is from
+   :n:`@ident` (if specified) or the construct, except that if the name from the
+   construct already exists in the :term:`local context`, Coq uses a fresh name
+   instead.  The constructs have these forms:
+   (See examples :ref:`here <intro_examples>`.)
 
-   .. tacv:: apply {+, @term with @bindings} in {+, @ident}
+   `forall x : T, …`
+     `x T` is a :term:`dependent premise`.  Removes `forall x`
+     from the goal and adds `x : T` to the context.
 
-      This does the same but uses the bindings to instantiate
-      parameters of :token:`term` (see :ref:`bindings`).
+   `A -> B`
+     `A` is a :term:`non-dependent premise`.  Removes `A ->` from
+     the goal and adds `H : A` to the context.
 
-   .. tacv:: eapply {+, @term {? with @bindings } } in {+, @ident}
+   `let x := c, …`
+     Removes `let x := c,` from the goal and adds `x := c : T`
+     to the context.
 
-      This works as :tacn:`apply … in` but turns unresolved bindings into
-      existential variables, if any, instead of failing.
+   .. todo PR
+     Need an example of an existential variable
```

Comment:
I guess not related to `intro`.  (I don't know/recall how you deal with `exists` goals.  For another day.)

Reply:
Oh, OK, but note that "existential variable" in Coq's terminology usually means `?X` (things that you can obtain after calling `eapply` for instance) and not a variable bound by an `exists`.

Reply:
That seems inconsistent with the mathematical notion of existential quantifiers.  A poor choice of terminology but unlikely we'll ever change it.

Reply:
It is actually semantically close to the notion of existential quantifiers since "existential variables" are things that we don't know yet but assume exist. Anyway, like you said, this is not a terminology that is ever likely to change.